### PR TITLE
fix: remove deleted vynil space root directory on gc

### LINKED ISSUE #8441: Dropped vinyl space leaves empty root LSM directory after GC
**Bug description**

When vinyl space is dropped, its files are left on the file system until GC removes them. At the moment GC removes only run files (`./521/0/00000000000000000046.run`, `./521/0/00000000000000000046.index` etc), but not the root directory (`./521/0` and `./521`). These empty directories are never removed and occupy 4KB on ext-family file systems each. In a case of many dropped vinyl spaces it can become a serious disk space and inode leak.

Current master branch of tarantool is affected.

**Steps to reproduce**

```lua
box.cfg{checkpoint_count = 1}

space_name = "test"
space = box.schema.space.create(space_name, {
  format = {
    { "id", "unsigned", is_nullable = false }
  },
  engine = "vinyl",
  if_not_exists = true
}) 

space:create_index("id", {
    type = "TREE",
    unique = true,
    if_not_exists = true,
    parts = { { field = "id", type = "unsigned" } },
  })

for i=1,10000 do space:insert({i}) end

box.snapshot()
    
space:drop()

box.snapshot()
```

**Actual behavior**

We can see that in the instance catolog the root directory still exists (here 522 - is the space id, and 0 - is the index id):
```
ls 522/0 
```

**Expected behavior**

No artifacts from the vinyl space left on the file system.

### PR DESCRIPTION
**Pull request by darthunix**
When vinyl space is dropped, its files are left on the file system until GC removes them. At the moment GC removes only run files, but not the root directory. These empty directories are never removed and occupy 4KB on ext-family file systems each. In a case of many dropped vinyl spaces it can become a serious disk space and inode leak. Current commit fixes this issue.

Closes #8441

NO_DOC=bugfix
NO_CHANGELOG=bugfix

### CODE REVIEW

File: test/vinyl-luatest/gh_8441_remove_root_directory_after_gc_test.lua [hunk 1]
```diff
@@ -0,0 +1,72 @@
+local server = require('luatest.server')
+local t = require('luatest')
+
+local g = t.group()
+
+g.before_all(function(cg)
+    cg.server = server:new({alias = 'master'})
+    cg.server:start()
+    cg.server:exec(function()
+        box.schema.space.create("test", {
```
Comment (locker) [inline test/vinyl-luatest/gh_8441_remove_root_directory_after_gc_test.lua:10]: Nit: Better create the test space right in the test so that it can be repeated. (Starting the server in `before_all` is okay though.) Currently the test fails with `luatest -r 2`:

<details>

```
$ luatest -v test/vinyl-luatest/gh_8441_remove_root_directory_after_gc_test.lua
Tarantool version is 3.0.0-entrypoint-69-g14e36a526ef9
Running with --shuffle group:2424
Started on Thu Mar 16 13:00:54 2023
    vinyl-luatest.gh_8441_remove_root_directory_after_gc.test_root_directory_is_removed_after_gc ... (0.401s) Ok
=========================================================
Ran 1 tests in 0.732 seconds, 1 succeeded, 0 failed
$ luatest -v test/vinyl-luatest/gh_8441_remove_root_directory_after_gc_test.lua -r 2
Tarantool version is 3.0.0-entrypoint-69-g14e36a526ef9
Running with --shuffle group:7072
Started on Thu Mar 16 13:00:56 2023
    vinyl-luatest.gh_8441_remove_root_directory_after_gc.test_root_directory_is_removed_after_gc ... (0.400s) error
...-luatest/gh_8441_remove_root_directory_after_gc_test.lua:44: attempt to index field 'test' (a nil value)
=========================================================

Tests with errors:
------------------

1) vinyl-luatest.gh_8441_remove_root_directory_after_gc.test_root_directory_is_removed_after_gc
...-luatest/gh_8441_remove_root_directory_after_gc_test.lua:44: attempt to index field 'test' (a nil value)
stack traceback:
        ...-luatest/gh_8441_remove_root_directory_after_gc_test.lua:38: in function 'vinyl-luatest.gh_8441_remove_root_directory_after_gc.test_root_directory_is_removed_after_gc'
        ...
        [C]: in function 'xpcall'

Ran 1 tests in 0.728 seconds, 0 succeeded, 1 errored

=========================================================
Failed tests:

vinyl-luatest.gh_8441_remove_root_directory_after_gc.test_root_directory_is_removed_after_gc
```

</details>

**Review by locker — CHANGES_REQUESTED** [review]

File: src/box/vinyl.c [hunk 2]
```diff
@@ -3225,6 +3264,9 @@ vy_gc(struct vy_env *env, struct vy_recovery *recovery,
 			if (loops % VY_YIELD_LOOPS == 0)
 				fiber_sleep(0);
 		}
+
+		if (lsm_needs_gc)
```
[the remark this answers is not in the document]
Comment (darthunix) [reply src/box/vinyl.c:3268]: `vy_run_remove_files()` is also called by vinyl scheduler on removing compacted run files. I don't think that it is a correct place to put the code that unlinks the space root directory (it is the case only for a fully deleted space).

File: src/box/vinyl.c [hunk 3]
```diff
@@ -3225,6 +3264,9 @@ vy_gc(struct vy_env *env, struct vy_recovery *recovery,
...
+			vy_gc_lsm_root(env, lsm_info);
```
[the remark this answers is not in the document]
Comment (darthunix) [reply src/box/vinyl.c:3269]: 1. I think that it is more a bug fix then a feature. The normal behavior for a deleted space is to remove all its files rather then to leave some of them polluting the file system. And current commit brings vinyl deletion to normal.
2. Added a changelog.
3. Fixed the typo, made commit header start with `vinyl: ...`.

File: test/vinyl-luatest/gh_8441_remove_root_directory_after_gc_test.lua (hunk 1 above)
Comment (darthunix) [reply test/vinyl-luatest/gh_8441_remove_root_directory_after_gc_test.lua:20]: Done.

File: src/box/vinyl.c [hunk 4]
```diff
@@ -3193,6 +3194,43 @@ vy_gc_lsm(struct vy_lsm_recovery_info *lsm_info)
...
 }
 
+/**
+ * Removes an empty directory.
+ *
+ * @param path Path to the directory.
+ * @return int 0 on success, -1 on error.
+ */
+static inline int
+rm_dir(const char *path)
```
[the remark this answers is not in the document]
Comment (darthunix) [reply src/box/vinyl.c]: Done.

File: test/vinyl-luatest/gh_8441_remove_root_directory_after_gc_test.lua [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+    cg.server:exec(function()
+        local fio = require('fio')
+
+        -- Make each snapshot trigger garbage collection.
+        box.cfg{checkpoint_count = 1}
+
+        local space_id = box.space.test.id
+
+        -- Populate runs with data.
+        for i=1,10000 do box.space.test:insert({i, i}) end
```
[the remark this answers is not in the document]
Comment (darthunix) [reply test/vinyl-luatest/gh_8441_remove_root_directory_after_gc_test.lua]: Done.

File: test/vinyl-luatest/gh_8441_remove_root_directory_after_gc_test.lua [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+        local r, e = fio.listdir(fio.pathjoin(box.cfg.vinyl_dir, space_id, 0))
+        t.assert_equals(r, nil)
+        t.assert_str_contains(tostring(e), expected_err_msg)
+
+        r, e = fio.listdir(fio.pathjoin(box.cfg.vinyl_dir, space_id, 1))
+        t.assert_equals(r, nil)
+        t.assert_str_contains(tostring(e), expected_err_msg)
+
+        r, e = fio.listdir(fio.pathjoin(box.cfg.vinyl_dir, space_id))
+        t.assert_equals(r, nil)
```
[the remark this answers is not in the document]
Comment (darthunix) [reply test/vinyl-luatest/gh_8441_remove_root_directory_after_gc_test.lua]: Done.

File: src/box/vinyl.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (locker) [reply src/box/vinyl.c:3268]: Why not? If there's no run files left after compaction, it looks natural to drop the directory as well, even if the space still exists. The directory will be created on demand, when a new run file is written (see `xlog_create` -> `mkdirpath`).

[the remark this answers is not in the document]
Comment (darthunix) [reply src/box/vinyl.c:3268]: 1. When we create a new vinyl space we produce root directory. The idea to remove any evidence of the vinyl space from the file system and then recover the directories while writing to WAL seems non-intuitive to me.
2. The name and the signature of the `vy_run_remove_files(const char *dir, uint32_t space_id, uint32_t iid, int64_t run_id)` function means that in must remove the files related to a specific run, not the root directory where all the runs live.
3. It is additional work to unlink/link root directory on compaction.

I still don't understand why this solution looks simpler then the current one.

[the remark this answers is not in the document]
Comment (locker) [reply src/box/vinyl.c:3268]: 1. The code that creates an index directory isn't required since commit c00ba8e76470e4dfe131220aaf03cb24c660f438. I think we can drop it now.
2. If a directory is empty, there's no point in keeping it so it can be removed by `vy_run_remove_files`.
3. This work is negligible in comparison to compaction.

> I still don't understand why this solution looks simpler then the current one.

Because it's difficult to comprehend how your solution actually works. Consider this example:

```Lua
box.cfg{checkpoint_count = 1}
box.schema.create_space('test', {engine = 'vinyl', id = 512})
box.space.test:create_index('pk')
box.space.test:insert{1}
box.snapshot()
box.space.test:drop()
box.schema.create_space('test', {engine = 'vinyl', id = 512})
box.space.test:create_index('pk')
box.snapshot()
```

Directory "512/0" will be dropped as a result of this script but space 512 exists.

I don't see any point in trying to describe or fix such corner cases while we can simply say that the space/index directory is removed as soon as there's no files left in it.

[the remark this answers is not in the document]
Comment (darthunix) [reply src/box/vinyl.c]: Ok, as far as you insist, here is  a new version. Now we always remove the root directory on GC (space deletion and compaction). I've preferred to keep functions simple and didn't change `vy_run_remove_files()`. Instead, a new `vy_root_remove_files()` function with similar signature was introduces. It is used to remove the empty directorie files now as far as I don't want to make `vy_run_remove_files()` overcomplicated.

[the remark this answers is not in the document]
Comment (locker) [reply src/box/vinyl.c]: Why would the directory removal code suddently make `vy_run_remove_files()` overcomplicated?

If directories were removed in `vy_run_remove_files()`, all the code dealing with file removal would be gathered in one place, not scattered around garbage collection and compaction scheduling.

[the remark this answers is not in the document]
Comment (darthunix) [reply src/box/vinyl.c]: Ok, this is the solution you want to see (if I understand you right). But I am not a big fan of trying to remove the directory for each run. All these attempts would fail (except the last one) and it seems to me a non-optimal code.

File: test/vinyl/layout.result [hunk 7]
```diff
@@ -185,16 +185,6 @@ result
...
-      - HEADER:
-          timestamp: <timestamp>
-          type: INSERT
-        BODY:
-          tuple: [7, {2: 5}]
-      - HEADER:
-          timestamp: <timestamp>
-          type: INSERT
-        BODY:
-          tuple: [7, {2: 4}]
```
Comment (darthunix) [inline test/vinyl/layout.result:197]: @locker I need your help here. I failed to understand the vylog internals myself (though I have read `rfc/vylog-ups-and-downs.md`) and I am not sure whether this change in the layout result is correct or not...

Comment (locker) [reply test/vinyl/layout.result:197]: This happens because `vy_run_remove_files()` now returns -1 if the index directory isn't empty. As a result, `vy_gc_run()` doesn't remove information about the run from the vylog, which leads to the vinyl/layout test failure.

`vy_run_remove_files()` should return 0 in this case.

File: src/box/vy_run.c [hunk 8]
```diff
@@ -2558,6 +2580,14 @@ vy_run_remove_files(const char *dir, uint32_t space_id,
 		} else
 			say_info("removed %s", path);
 	}
+
+	/* Remove the root directory if it's empty. */
+	vy_lsm_snprint_path(path, sizeof(path), dir, space_id, iid);
+	if (try_rmdir(path) < 0)
+		return -1;
+	snprintf(path, sizeof(path), "%s/%u", dir, space_id);
```
Comment (locker) [inline src/box/vy_run.c:2588]: Please add a helper function `vy_space_snprint_path()` and make `vy_lsm_snprint_path()` use it so that path formatting remains in one place.

File: src/box/vinyl.c [hunk 9]
```diff
@@ -3195,6 +3196,7 @@ vy_gc_lsm(struct vy_lsm_recovery_info *lsm_info)
 
 /**
  * Delete unused run files stored in the recovery context.
+ * If there are no more runs, delete the root directory as well.
```
Comment (locker) [inline src/box/vinyl.c:3199]: Changes to this file are left from the previous patch version. Please revert them since they aren't needed anymore.

Comment (darthunix) [reply src/box/vinyl.c]: Fixed.

File: src/box/vy_run.c (hunk 8 above)
Comment (darthunix) [reply src/box/vy_run.c]: Fixed.

File: test/vinyl/layout.result (hunk 7 above)
Comment (darthunix) [reply test/vinyl/layout.result]: Fixed.

**Comment by locker** [comment]
Cherry-picked to 2.11.
